### DISCUSSION TITLE
Should I use Server Actions in Next.js instead of API Route Handlers for fetching data?

### DISCUSSION THREAD

**Discussion by TimoP** [vercel/next.js#72919] [Help] [answered]
### Summary

Hi everyone,

I'm relatively new to Next.js and recently started working with the App Router. Initially, I fetched data from the server using Route Handlers (API routes). However, I discovered that I can directly call Server Actions from components, such as within a useEffect hook or event handlers, to fetch the same data.

This approach seems to simplify my codebase since I no longer need to write multiple Route Handlers for fetching data. Instead, I can directly call the Server Action and handle everything in one place.

My initial thought is that using Server Actions reduces boilerplate and keeps my code cleaner. However, I'm unsure if this is a good practice in the long term.

### Additional information

The both ways. Imagine other fancy client code in the TestComponent.

```js
// Using API Route Handlers

const TestComponent = () => {
const [niceData, setData] = useState([]);

useEffect(() => {
  fetch('api/data')
    .then((r) => r.json())
    .then((data) => setData(data))
}, []);

return (<div>{data.map(d => <div>{d}</div>)}</div>)
}

// /api/data/route.ts
export const GET = async (request) => {
  // Stuff to get the data from DB
  return new Response(JSON.stringify({ data }), { status: 200 });
}
```
```js
// VS Using Server Actions

import { getData } from './serverActions';

const TestComponent = () => {
const [niceData, setData] = useState([]);

useEffect(() => {
  getData()
    .then((data) => setData(data))
}, []);

return (<div>{data.map(d => <div>{d}</div>)}</div>)
}

// serverActions.ts
export const getData = async () => {
  // Stuff to get the data from DB
  return data;
}
```

**Comment by stijnwesterhof** [comment]
Since there is no state to be managed within this component, I assume that this could just be a server component. In that case you don't have to use `useEffect` to do the data fetching.

```
const TestComponent = async () => {
  const data = await getData();

  return (
    <div>
      {data.map((d) => (
        <div>{d}</div>
      ))}
    </div>
  );
};

**Reply by TimoP** [reply]
its only demo code to show both ways. Imagine fancy client stuff in the components :)

**Reply by stijnwesterhof** [reply]
Since you stated that you're relatively new to Next.js, I thought I could help you by suggesting to use a server component instead :) 

It's also what what the docs on https://nextjs.org/docs/pages/building-your-application/routing/api-routes recommends you to do. 
> Good to know: If you are using the App Router, you can use [Server Components](https://nextjs.org/docs/app/building-your-application/rendering/server-components) or [Route Handlers](https://nextjs.org/docs/app/building-your-application/routing/route-handlers) instead of API Routes.

This probably answers your question.

**Reply by TimoP** [reply]
Thank you. The `fetch('api/data')` would call a Route Handler.
My question is whether it is good practice to "fetch" data directly with the Server Action in a client component (maybe deeper in the tree) instead of fetching the data with a Route Handler and a fetch call. 

If the parent is a server component, I will use a prop to get the data in the client component.
I have updated my examples to make the two approaches clearer.

For me, at the moment, the direct call of the Server Action from the effect hook is the more direct way with less overhead in comparison. But maybe I miss a few downsides from that approach. In the end that's the question, are there downsides and do I have to dive deeper into that aspect of nextjs.

**Reply by icyJoseph** [reply, marked as answer]
- https://react.dev/reference/rsc/use-server#caveats

> Server Actions are designed for mutations that update server-side state; they are not recommended for data fetching. Accordingly, frameworks implementing Server Actions typically process one action at a time and do not have a way to cache the return value.

**Reply by TimoP** [reply]
@icyJoseph Thank you, thats helpful.
